src/Aspire.Dashboard/Components/Controls/ResourceDetails.razor.cs

```diff
@@ -117,111 +113,101 @@ private IEnumerable<DisplayedEndpoint> GetEndpoints()
-                k => k.Key,
-                (p, k) => new SummaryValue { Key = p.Key, Value = p.Value, KnownProperty = k.SingleOrDefault(), Tooltip = GetTooltip(p.Value) })
-            .OrderBy(v => v.KnownProperty != null ? resolvedKnownProperties.IndexOf(v.KnownProperty) : int.MaxValue);
-
-        return values;
-    }
+        var vms = Resource.Properties.Values
+            .Where(vm => vm.Value is { HasNullValue: false } and not { KindCase: Value.KindOneofCase.ListValue, ListValue.Values.Count: 0 });
 
-    private static string GetTooltip(Value value)
-    {
-        if (value.HasStringValue)
-        {
-            return value.StringValue;
-        }
-        else
-        {
-            return value.ToString();
-        }
-    }
+        return ordered
+            ? vms.OrderBy(vm => vm.Priority).ThenBy(vm => vm.Name)
+            : vms;
 
-    private static string GetDisplayedValue(BrowserTimeProvider timeProvider, SummaryValue summaryValue)
-    {
-        string value;
-        if (summaryValue.Value.HasStringValue)
-        {
-            value = summaryValue.Value.StringValue;
-        }
-        else
-        {
-            // Complex values such as arrays and objects will be output as JSON.
-            // Consider how complex values are rendered in the future.
-            value = summaryValue.Value.ToString();
-        }
-        if (summaryValue.Key == KnownProperties.Container.Id)
-        {
-            // Container images have a short ID of 12 characters
-            value = value.Substring(0, Math.Min(value.Length, 12));
-        }
-        else
+        void PopulateMissingKnownProperties()
         {
-            // Dates are returned as ISO 8601 text.
-            // Use try parse to check if a value matches ISO 8601 format. If there is a match then convert to a friendly format.
-            if (DateTime.TryParseExact(value, "o", CultureInfo.InvariantCulture, DateTimeStyles.None, out var date))
+            List<KnownProperty>? knownProperties = null;
+
+            // Lazily populate some data on the view models, that wasn't available when they were constructed.
+            foreach ((var key, var vm) in Resource.Properties)
             {
-                value = FormatHelpers.FormatDateTime(timeProvider, date, cultureInfo: CultureInfo.CurrentCulture);
+                // A priority of -1 means the known property hasn't been looked up. Do that once per property, now.
+                if (vm.Priority == -1)
+                {
+                    knownProperties ??= GetKnownPropertiesForSelectedResourceType();
+
+                    var found = false;
+
+                    for (var i = 0; i < knownProperties.Count; i++)
+                    {
+                        var kp = knownProperties[i];
+
+                        if (kp.Key == key)
+                        {
+                            found = true;
+                            vm.KnownProperty = kp;
+                            vm.Priority = i;
```

Comment:
We store a priority directly on the object to avoid the _O(N)_ `IndexOf` within an _O(N log N)_ `OrderBy`, leading to _O(N^2 log N)_ which is quasi-quadratic rather than log-linear time complexity.

We also only do this lookup once, rather than on every render of the UI. So after this, it's constant time.